pyro/distributions/transforms/naf.py

```diff
@@ -13,12 +13,123 @@
 eps = 1e-8
 
 
+class ELUMixin(object):
+    @staticmethod
+    def f(x):
+        """
+        Implements the nonlinearity of NAF, in this case ELU
+        """
+        return F.elu(x)
+
+    @staticmethod
+    def f_inv(x):
+        """
+        Implements the inverse of ELU
+        """
+        return torch.max(x, torch.zeros_like(x)) + torch.min(torch.log1p(x + eps), torch.zeros_like(x))
+
+    @staticmethod
+    def log_df_dx(x):
+        """
+        Implements the log derivative of NAF nonlinearity
+        """
+        return -F.relu(-x)
+
+    @staticmethod
+    def log_df_inv_dx(x):
+        """
+        Implements the log derivative of inverse NAF nonlinearity
+        """
+        return F.relu(-torch.log1p(x + eps))
+
+
+class LeakyReLUMixin(object):
+    @staticmethod
+    def f(x):
+        """
+        Implements the nonlinearity of NAF, in this case leaky ReLU
+        """
+        return F.leaky_relu(x)
+
+    @staticmethod
+    def f_inv(x):
+        """
+        Implements the inverse of leaky ReLU
+        """
+        # slope for negative part is inverse of slope for positive part in f(x)
+        return F.leaky_relu(x, negative_slope=100.0)
+
+    @staticmethod
+    def log_df_dx(x):
+        """
+        Implements the log derivative of NAF nonlinearity
+        """
+        return torch.where(x >= 0., torch.zeros_like(x), torch.ones_like(x) * math.log(0.01))
+
+    @staticmethod
+    def log_df_inv_dx(x):
+        """
+        Implements the log derivative of inverse NAF nonlinearity
+        """
+        return torch.where(x >= 0., torch.zeros_like(x), torch.ones_like(x) * math.log(100.0))
+
+
+def safe_log(x):
+    return torch.log(x * 1e2) - math.log(1e2)
+
+
+class SigmoidalMixin(object):
+    @staticmethod
+    def f(x):
+        """
+        Implements the nonlinearity of NAF, in this case sigmoid with scaled output
+        """
+        return torch.sigmoid(x) * (1. - eps) + 0.5 * eps
+
+    @staticmethod
+    def f_inv(x):
+        """
+        Implements the inverse scaled sigmoid nonlinearity
+        """
+        y = (x - 0.5 * eps) / (1. - eps)
+        return safe_log(y) - safe_log(1. - y)
```

Comment:
Instead of `safe_log(1. - y)`, could you instead use `log1p(-y)`?

Reply:
`safe_log` is a way to get extra precision at the lower end of the input. I copied it from the author's implementation, but I'm not sure whether it's entirely necessary. `log1p(-y)` would work instead of `safe_log(1. - y)` and I'll change that, but I don't think `log1p(y-1)` would adequately replace `safe_log(y)`. 

It's perhaps better to keep it in just in case?

Reply:
> log(x) where x is almost 0

I think log(x) should have good precision when x is quite small without resorting to this, and log1p can be used for values close to 1. We are using this function at quite a few places, so I'll leave it to you to decide whether to keep this! :)

Reply:
I've removed `safe_log`